### PR TITLE
[IPO-202] Add initial Data Collector application and /_status check

### PR DESCRIPTION
**Pull request by ryancragun**
The Data Collector application is intended to export Chef Server actions
to an external Data Collector server via HTTP requests. This PR adds:

 * The initial Data Collector application which consists of a supervisor
   and an HTTP connection pool.
 * Status checks callbacks for oc_chef_wm that are used application has
   been enabled.

### CODE REVIEW

File: omnibus/files/private-chef-cookbooks/private-chef/libraries/private_chef.rb [hunk 1]
```diff
@@ -89,6 +89,21 @@ module PrivateChef
   backup Mash.new
   backup["strategy"] = "tar"
 
+  #
```
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/libraries/private_chef.rb]: This comment might be more useful in the attributes file, that is where most of the other similar comments are.

File: src/oc_erchef/apps/data_collector/src/data_collector_http.erl [hunk 2]
```diff
@@ -0,0 +1,85 @@
+-module(data_collector_http).
```
Comment (stevendanna) [inline src/oc_erchef/apps/data_collector/src/data_collector_http.erl]: It it mildly troublesome that we now have this same basic code in 4-5 different places (across 3-4 projects).  I think it points to us wanting a different API from opscoderl_httpc.

File: src/oc_erchef/apps/data_collector/src/data_collector_http.erl [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            error_logger:error_report({?MODULE, Method, How, Why}),
+            {error, Why}
+    end.
+
+%%
+%% Pool management functions
+%%
+-spec create_pool() -> ok.
+create_pool() ->
+    error_logger:info_msg("Creating Data Collector HTTP pool.~n"),
```
Comment (stevendanna) [inline src/oc_erchef/apps/data_collector/src/data_collector_http.erl]: Can we use lager for the logging in this module instead?

File: src/oc_erchef/apps/data_collector/src/data_collector_app.erl [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+-behaviour(application).
+
+%% API
+-export([start/2,
+         stop/1
+        ]).
+
+start(_StartType, _StartArgs) ->
+    error_logger:info_msg("Starting Data Collector app.~n"),
+    supervisor:start_link({local, data_collector_sup}, data_collector_sup, []).
```
Comment (stevendanna) [inline src/oc_erchef/apps/data_collector/src/data_collector_app.erl]: Minor point: I think that in our other applications we often expose a start_link/0 in the supervisor that calls supervisor:start_link

File: src/oc_erchef/apps/chef_index/src/chef_index_http.erl [hunk 5]
```diff
@@ -31,43 +31,43 @@ request(Path, Method, Body, Headers) ->
 %% care about success or failure.
 %%
 -spec get(list()) -> ok | {error, term()}.
-get(Url) ->
-    get(Url, [], ?DEFAULT_HEADERS).
+get(Path) ->
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_index/src/chef_index_http.erl]: These changes don't appear related to this PR?

File: src/oc_erchef/apps/data_collector/.gitignore [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+ebin_dialyzer/
+TAGS
+.DS_Store
+doc/*.html
+*.beam
+/doc/edoc-info
+/doc/erlang.png
+/doc/stylesheet.css
+/deps.plt
+/.concrete
```
Comment (marcparadise) [inline src/oc_erchef/apps/data_collector/.gitignore]: I don't think we're using concrete anymore?

**Comment by marcparadise** [comment]
Looks like this is on the right track. I'd like to understand better the background of data collector, and how it'll differ from the capture we are currently doing for analytics.

It would also be helpful  to separate out the changes that are formatting/indent/naming-related into their own  PR.

**Comment by ryancragun** [comment]
I'll rebase this to remove the first style commit as soon as https://github.com/chef/chef-server/pull/860 has been merged

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb [hunk 7]
```diff
@@ -2,304 +2,351 @@
...
+      {init_count, <%= @solr_http_init_count %>},
+      {max_count, <%= @solr_http_max_count %>},
+      {cull_interval, <%= @solr_http_cull_interval %>},
+      {max_age, <%= @solr_http_max_age %>},
+      {max_connection_duration, <%= @solr_http_max_connection_duration %>},
+      {ibrowse_options, <%= @solr_ibrowse_options %>}
+    ]}
+  ]},
+
+  <% if node['private_chef']['data_collector']['root_url'] && node['private_chef']['data_collector']['token'] %>
```
Comment (adamleff) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb]: Per the RFC, the token is optional. Chef's commercial product that relies on Data Collector requires the token, but others' implementations may not.

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb [hunk 8]
```diff
@@ -2,304 +2,351 @@
...
+  {data_collector, [
+    {root_url,                "<%= node['private_chef']['data_collector']['root_url'] %>"},
+    {token,                   <%= node['private_chef']['data_collector']['token'] %>},
```
Comment (adamleff) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb]: Token may be alphanumeric - we never settled on an exact spec. Should this be in quotes?

File: src/oc_erchef/apps/data_collector/src/data_collector_http.erl [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+-spec delete(list(), iolist() | binary(), list()) -> ok | {error, term()}.
+delete(Path, Body, Headers) ->
+    request_with_caught_errors(Path, delete, Body, Headers).
+
+request_with_caught_errors(Path, Method, Body, Headers) when is_list(Body)->
+    request_with_caught_errors(Path, Method, iolist_to_binary(Body), Headers);
+request_with_caught_errors(Path, Method, Body, Headers) ->
+    try
+        case request(Path, Method, Body, Headers) of
+            {ok, "200", _Head, _RespBody} -> ok;
```
Comment (adamleff) [inline src/oc_erchef/apps/data_collector/src/data_collector_http.erl]: Chef's implementation of the Data Collector server currently returns a `204 No Content` when we post the payload. We probably need to account for that and perhaps consider any 2xx status code as OK.

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb [hunk 10]
```diff
@@ -2,304 +2,351 @@
...
+    {actions_vhost, <<"<%= @actions_vhost %>">>},
+    {actions_exchange, <<"<%= @actions_exchange %>">>},
+    {actions_fqdn, <<"<%= node['private_chef']['lb']['api_fqdn'] %>">>},
+    {max_request_size, <%= node['private_chef']['opscode-erchef']['max_request_size'] %>},
+    {server_version, "<%= node['private_chef']['version'] %>"},
+    {health_ping_timeout, <%= node['private_chef']['opscode-erchef']['health_ping_timeout'] %>},
+    {health_ping_modules, [
+        oc_chef_authz,
+        chef_sql,
+        <% if node['private_chef']['data_collector']['root_url'] && node['private_chef']['data_collector']['token'] %>
```
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb]: Similar issue with `&& token` here that adamleff mentioned below.

**Comment by ryancragun** [comment]
http://wilson.ci.chef.co/job/chef-server-12-trigger-ad_hoc/187/downstreambuildview/

File: omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb [hunk 11]
```diff
@@ -217,30 +217,19 @@
         ]}
     ]},
 
-    {chef_objects, [
```
Comment (stevendanna) [inline omnibus/files/private-chef-cookbooks/private-chef/templates/default/oc_erchef.config.erb]: Why is this being deleted?

File: src/oc_erchef/apps/data_collector/src/data_collector.erl [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    end.
+
+-spec update(iolist() | binary()) -> ok | {error, term()}.
+update(Body) when is_list(Body) ->
+    update(iolist_to_binary(Body));
+update(Body) ->
+    %% TODO: Transform to data collector JSON
+    data_collector_http:post("/", Body).
+
+-spec is_enabled() -> true | false.
```
Comment (stevendanna) [inline src/oc_erchef/apps/data_collector/src/data_collector.erl]: Minor: You can use the boolean() type for the return here.

File: src/oc_erchef/apps/data_collector/src/data_collector_http.erl (hunk 9 above)
Comment (stevendanna) [inline src/oc_erchef/apps/data_collector/src/data_collector_http.erl]: Maybe match on the list `[$2|_Rest]`?

**Comment by stevendanna** [comment]
A couple of comments, but overall this looks fine.  I assume we'll add tests for update/1 once that is called from the application and more than a small stub method.

File: src/oc_erchef/apps/data_collector/src/data_collector.erl [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    case data_collector_http:get("/") of
+        ok -> pong;
+        _ -> pang
+    end.
+
+-spec update(iolist() | binary()) -> ok | {error, term()}.
+update(Body) when is_list(Body) ->
+    update(iolist_to_binary(Body));
+update(Body) ->
+    %% TODO: Transform to data collector JSON
```
Comment (marcparadise) [inline src/oc_erchef/apps/data_collector/src/data_collector.erl]: Can we track this with a card instead of a source TODO?

**Comment by marcparadise** [comment]
:+1: once the deleted config is restored

File: src/oc_erchef/apps/data_collector/src/data_collector.erl (hunk 13 above)
Comment (ryancragun) [inline src/oc_erchef/apps/data_collector/src/data_collector.erl]: Yup, IPO-203 is for that implementation.
